dask/dataframe/io/io.py

```diff
@@ -454,48 +454,59 @@ def from_dask_array(x, columns=None, index=None, meta=None):
     """
     meta = _meta_from_array(x, columns, index, meta=meta)
 
-    if x.ndim == 2 and len(x.chunks[1]) > 1:
-        x = x.rechunk({1: x.shape[1]})
+    name = "from-dask-array-" + tokenize(x, columns)
+    graph_dependencies = [x]
+    arrays_and_indices = [x.name, "ij" if x.ndim == 2 else "i"]
+    numblocks = {x.name: x.numblocks}
 
-    name = "from-dask-array" + tokenize(x, columns)
-    to_merge = []
+    if is_series_like(meta):
+        kwargs = {"dtype": x.dtype, "name": meta.name}
+    else:
+        kwargs = {"columns": meta.columns}
 
     if index is not None:
-        if not isinstance(index, Index):
-            raise ValueError("'index' must be an instance of dask.dataframe.Index")
         if index.npartitions != x.numblocks[0]:
             msg = (
                 "The index and array have different numbers of blocks. "
                 "({} != {})".format(index.npartitions, x.numblocks[0])
             )
             raise ValueError(msg)
         divisions = index.divisions
-        to_merge.append(ensure_dict(index.dask))
-        index = index.__dask_keys__()
+        graph_dependencies.append(index)
+        arrays_and_indices.extend([index.__dask_layers__()[0], "i"])
+        numblocks[index.__dask_layers__()[0]] = (index.npartitions,)
 
     elif np.isnan(sum(x.shape)):
         divisions = [None] * (len(x.chunks[0]) + 1)
-        index = [None] * len(x.chunks[0])
     else:
+        dsk = {}
         divisions = [0]
-        for c in x.chunks[0]:
-            divisions.append(divisions[-1] + c)
-        index = [
-            (np.arange, a, b, 1, "i8") for a, b in zip(divisions[:-1], divisions[1:])
-        ]
+        stop = 0
+        index_name = name + "-index-" + tokenize(x.chunks[0])
+        for i, increment in enumerate(x.chunks[0]):
+            stop += increment
+            dsk[(index_name, i)] = (pd.RangeIndex, divisions[i], stop)
+            divisions.append(stop)
         divisions[-1] -= 1
+        index = new_dd_object(dsk, index_name, pd.RangeIndex(0, 1), divisions)
```

Comment:
I still think a `BlockwiseDep` would be a simpler way to do this, and result in a cleaner, fused blockwise graph (the code would look pretty similar, too!)

Reply:
What about something kinda like
```python
slices = ArraySliceDep(x.chunks[:1])
index = map_partitions(lambda s: pd.RangeIndex(s.start, s.stop), slices, meta=pd.RangeIndex)
```
This isn't quite right
1. Shouldn't use a lambda because of pickling overhead
2. meta probably doesn't work
3. @ian-r-rose doesn't like using `map_partitions` this way :)

But basically, make the index via a blockwise operation using an `ArraySliceDep` or similar.

Reply:
I certainly am okay with this use of `ArraySliceDep`! I might just feed it directly into the one `blockwise` invocation rather than creating the `index` separately, though :)

Reply:
@ian-r-rose - Perhaps this is similar to the from_delayed problem?  It seems like we want something like an IO dep (i.e. a `BlockwiseDep`), but we want the Blockwise logic to recognize that these "special" IO deps are" real" (task or delayed) dependencies.

Reply:
I think it shares some features in common, but as you note below, this is probably doable with a local range index within the same task.

Reply:
> this is probably doable with a local range index within the same task.

This is certainly true, but my BlockwiseDep comment is mostly refering to a desire to avoid using `blockwise`/`Blockwise` explicitly.  However, I'm now realizing the key distinction between array-collection keys and delayed keys is that the array-collection keys can already be specified with `indices` (while a general delayed key may or may not be a tuple).